src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java

```diff
@@ -224,15 +227,142 @@ private boolean learnBackjumpAddFromConflict(ConflictCause conflictCause) {
+		// Justify one MBT assigned atom.
+		Integer toJustify = ((ArrayAssignment) assignment).getSomeMBTAssignedAtom();
+		if (LOGGER.isDebugEnabled()) {
+			LOGGER.debug("Searching for justification of " + toJustify + " / " + atomTranslator.atomToString(atomOf(toJustify)));
+			LOGGER.debug("Assignment is (TRUE part only):" + translate(assignment.getTrueAssignments()));
+		}
+		Set<Literal> reasonsForUnjustified = analyzingGrounder.justifyAtom(toJustify, assignment);
+
+		// Turn the justification into a NoGood.
+		int[] reasons = new int[reasonsForUnjustified.size() + 1];
+		reasons[0] = toJustify;
+		int arrpos = 1;
+		for (Literal literal : reasonsForUnjustified) {
+			reasons[arrpos++] = (literal.isNegated() ? -1 : 1) * atomTranslator.get(literal.getAtom());
+		}
+		NoGood noGood = new NoGood(reasons);
+		int noGoodID = grounder.register(noGood);
+		Map<Integer, NoGood> obtained = new LinkedHashMap<>();
+		obtained.put(noGoodID, noGood);
+		LOGGER.debug("Learned NoGood is: " + atomTranslator.noGoodToString(noGood));
+		// Add NoGood and trigger backjumping.
+		if (!ingest(obtained)) {
+			logStats();
+			return false;
+		}
+		return true;
+	}
+
+	private boolean treatConflictAfterClosing(NoGood violatedNoGood) {
+		if (disableJustificationAfterClosing || disableJustifications || !(grounder instanceof ProgramAnalyzingGrounder)) {
+			// Will not learn from violated NoGood, do simple backtrack.
+			LOGGER.debug("NoGood was violated after all unassigned atoms were assigned to false; will not learn from it; skipping.");
+			if (!backtrack()) {
+				logStats();
+				return false;
+			}
+			return true;
+		}
+		ProgramAnalyzingGrounder analyzingGrounder = (ProgramAnalyzingGrounder) grounder;
+		LOGGER.debug("Justifying atoms in violated nogood.");
+		LinkedHashSet<Integer> toJustify = new LinkedHashSet<>();
+		// Find those literals in violatedNoGood that were just assigned false.
+		for (Integer literal : violatedNoGood) {
+			if (unassignedAtoms.contains(atomOf(literal))) {
+				toJustify.add(literal);
+			}
+		}
+		// Since the violatedNoGood may contain atoms other than BasicAtom, these have to be treated.
+		Map<Integer, NoGood> obtained = new LinkedHashMap<>();
+		Iterator<Integer> toJustifyIterator = toJustify.iterator();
+		ArrayList<Integer> ruleAtomReplacements = new ArrayList<>();
+		while (toJustifyIterator.hasNext()) {
+			Integer literal = toJustifyIterator.next();
+			Atom atom = atomTranslator.get(atomOf(literal));
+			if (atom instanceof BasicAtom) {
+				continue;
+			}
+			if (!(atom instanceof RuleAtom)) {
+				// Ignore atoms other than RuleAtom.
+				toJustifyIterator.remove();
+				continue;
+			}
+			// For RuleAtoms in toJustify the corresponding ground body contains BasicAtoms that have been assigned FALSE in the closing.
+			// First, translate RuleAtom back to NonGroundRule + Substitution.
+			String ruleId = (String) ((ConstantTerm)atom.getTerms().get(0)).getObject();
```

Comment:
ConstantTerm is a raw type. References to generic type ConstantTerm<T> should be parameterized

Reply:
Well, since this is a cast, doing so will result in an unchecked cast warning from the compiler. Due to type erasure there is no way for the runtime environment to be sure that the object of the ConstantTerm is a String. As far as I know there are only two options: using a generic cast and suppressing the unchecked cast warning, or doing it like above. The latter option seems preferable to me.

Reply:
The generic cast does _not_ produce a warning:
`String ruleId = (String) ((ConstantTerm<?>)atom.getTerms().get(0)).getObject();`
This option seems preferable to me because Eclipse IDE does not produce any warning for it (in contrast to the option currently implemented).